core/src/main/scala/chisel3/Probe.scala

```diff
@@ -0,0 +1,110 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3
+
+import chisel3._
+import chisel3.Data.ProbeInfo
+import chisel3.reflect.DataMirror.isProbe
+import chisel3.internal.{requireIsChiselType, Builder}
+import chisel3.internal.Builder.pushCommand
+import chisel3.internal.sourceinfo.SourceInfo
+import chisel3.internal.firrtl._
+
+/** Utilities for creating and working with probe types.
+  */
+object Probe {
+
+  private def apply_impl[T <: Data](source: => T, writable: Boolean): T = {
+    val prevId = Builder.idGen.value
+    val data = source // should only evaluate source once
+    requireIsChiselType(data)
+    val ret = if (!data.mustClone(prevId)) data else data.cloneType.asInstanceOf[T]
+    ret.probeInfo = Some(ProbeInfo(writable))
+    ret
+  }
+
+  /** Create a read-only probe type from a Chisel type. This is only used for
+    * ports and not for hardware.
+    */
```

Comment:
```suggestion
  /** Create a read-only probe type from a Chisel type.
    *
    * This is only used for creating ports which may be lowered to cross-module references
    */
```

Maybe? I dont know what you mean by "ports and not for hardware", before this PR ports *are* hardware?